File: LeetCode/Longest_Substring_Without_Repeating_Characters.java
/**
 * leetcode problem link : https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * Runtime : 6ms faster than 80% submissions
 * Memory : 39.7MB better than 74%
 */

import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) 
    {
        if(s.length()==1)
            return 1;
        HashSet set = new HashSet();
        
        int max=0;
        int l=0;
        int r=0;
        
        while(r<s.length())
        {
                        
            char c = s.charAt(r);
            
            if(! set.add(c))
            {
                max=java.lang.Math.max(max, set.size());
                
                while(s.charAt(l)!=c)
                {
                    set.remove(s.charAt(l++));
                }                
                set.remove(s.charAt(l++));
            }
            else
                r++;
            
        }
        
        max=java.lang.Math.max(max, set.size());
        return max;
    }
}

File: LeetCode/longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right = 0, 0
        count = {}
        max_len = 0
        # max_ss = None
        while right < len(s):
            if s[right] in count:
                count[s[right]] += 1
                while left<= right and count[s[right]] > 1:
                    # if right-left >= max_len:
                    #     max_ss = s[left:right]
                    count[s[left]] -=1
                    left += 1
                right += 1
            else:
                count[s[right]] = 1
                right += 1
                # if right-left >= max_len:
                #     max_ss = s[left:right]

            max_len = max(max_len, right-left)
                    
        return max_len
            
        